File: SimpleCADAPI/src/simplecadapi/scene/documents.py
"""Immutable aggregate boundaries for validated Scene contract documents."""

from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from types import MappingProxyType
from typing import Any, Callable, ClassVar, Generic, Mapping, TypeVar, cast

from typing_extensions import Self

from .canonical import canonical_json_bytes, parse_canonical_json
from .generated.connector_binding import (
    ConnectorBindingDocument as ConnectorBindingValue,
)
from .generated.entities import EntityDocument as EntityValue
from .generated.normalized_product import (
    NormalizedProductDocument as NormalizedProductValue,
)
from .generated.presentation import PresentationDocument as PresentationValue
from .generated.scene import SceneDocument as SceneValue
from .validation import (
    SceneContractError,
    SceneValidationReport,
    validate_connector_binding,
    validate_entity_asset,
    validate_normalized_product,
    validate_presentation,
    validate_scene_manifest,
)


DocumentValue = TypeVar("DocumentValue")


def _copy_json_value(value: Any, stack: set[int] | None = None) -> Any:
    """Detach JSON-like input while accepting frozen mappings and tuples."""

    if value is None or isinstance(value, (bool, int, float, str)):
        return value
    if stack is None:
        stack = set()
    if isinstance(value, Mapping):
        marker = id(value)
        if marker in stack:
            raise ValueError("cyclic value is not JSON")
        stack.add(marker)
        try:
            result: dict[str, Any] = {}
            for key, child in value.items():
                if not isinstance(key, str):
                    raise TypeError("JSON object keys must be strings")
                result[key] = _copy_json_value(child, stack)
            return result
        finally:
            stack.remove(marker)
    if isinstance(value, (list, tuple)):
        marker = id(value)
        if marker in stack:
            raise ValueError("cyclic value is not JSON")
        stack.add(marker)
        try:
            return [_copy_json_value(child, stack) for child in value]
        finally:
            stack.remove(marker)
    raise TypeError(f"unsupported JSON value: {type(value).__name__}")


def _freeze_json_value(value: Any) -> Any:
    if isinstance(value, dict):
        return MappingProxyType(
            {key: _freeze_json_value(child) for key, child in value.items()}
        )
    if isinstance(value, list):
        return tuple(_freeze_json_value(child) for child in value)
    return value


@dataclass(frozen=True, slots=True, init=False)
class _ContractDocument(Generic[DocumentValue]):
    _value: Mapping[str, Any] = field(repr=False, compare=False)
    _canonical_bytes: bytes = field(repr=False)
    _canonical_hash: str

    _validator: ClassVar[Callable[[Any], SceneValidationReport]]

    def __init__(self, value: DocumentValue | Mapping[str, Any]) -> None:
        copied = _copy_json_value(value)
        canonical = canonical_json_bytes(copied)
        parsed = parse_canonical_json(canonical)
        report = type(self)._validator(parsed)
        if not report.valid:
            raise SceneContractError(report)
        if not isinstance(parsed, dict):  # The schemas require object roots.
            raise TypeError("contract document root must be a JSON object")
        object.__setattr__(self, "_value", _freeze_json_value(parsed))
        object.__setattr__(self, "_canonical_bytes", canonical)
        object.__setattr__(
            self,
            "_canonical_hash",
            "sha256:" + hashlib.sha256(canonical).hexdigest(),
        )

    @classmethod
    def from_value(cls, value: DocumentValue | Mapping[str, Any]) -> Self:
        """Validate and detach a structural contract value."""

        return cls(value)

    @classmethod
    def parse(cls, data: bytes | bytearray | memoryview | str) -> Self:
        """Parse exact canonical JSON bytes, validate them, and freeze the value."""

        return cls(parse_canonical_json(data))

    @property
    def value(self) -> Mapping[str, Any]:
        """The deeply immutable document value."""

        return self._value

    @property
    def canonical_bytes(self) -> bytes:
        return self._canonical_bytes

    @property
    def canonical_hash(self) -> str:
        return self._canonical_hash

    def to_mutable(self) -> DocumentValue:
        """Return a fresh mutable dict/list tree with no aliases to this document."""

        return cast(DocumentValue, parse_canonical_json(self._canonical_bytes))

    def __bytes__(self) -> bytes:
        return self._canonical_bytes


@dataclass(frozen=True, slots=True, init=False, eq=False, repr=False)
class SceneDocument(_ContractDocument[SceneValue]):
    _validator: ClassVar[Callable[[Any], SceneValidationReport]] = (
        validate_scene_manifest
    )

    @property
    def revision(self) -> str:
        """The validated two-pass revision carried by this immutable scene."""

        return cast(str, self._value["revision"])


@dataclass(frozen=True, slots=True, init=False, eq=False, repr=False)
class EntityDocument(_ContractDocument[EntityValue]):
    _validator: ClassVar[Callable[[Any], SceneValidationReport]] = (
        validate_entity_asset
    )


@dataclass(frozen=True, slots=True, init=False, eq=False, repr=False)
class PresentationDocument(_ContractDocument[PresentationValue]):
    _validator: ClassVar[Callable[[Any], SceneValidationReport]] = (
        validate_presentation
    )


@dataclass(frozen=True, slots=True, init=False, eq=False, repr=False)
class ConnectorBindingDocument(_ContractDocument[ConnectorBindingValue]):
    _validator: ClassVar[Callable[[Any], SceneValidationReport]] = (
        validate_connector_binding
    )


@dataclass(frozen=True, slots=True, init=False, eq=False, repr=False)
class NormalizedProductDocument(_ContractDocument[NormalizedProductValue]):
    _validator: ClassVar[Callable[[Any], SceneValidationReport]] = (
        validate_normalized_product
    )


__all__ = [
    "ConnectorBindingDocument",
    "EntityDocument",
    "NormalizedProductDocument",
    "PresentationDocument",
    "SceneDocument",
]

File: SimpleCADAPI/scene-contract/src/documents.ts
import type {
  ConnectorBindingDocument as ConnectorBindingValue,
  EntityDocument as EntityValue,
  NormalizedProductDocument as NormalizedProductValue,
  PresentationDocument as PresentationValue,
  SceneDocument as SceneValue,
} from "./generated/index.js";
import { canonicalJsonBytes, parseCanonicalJson, sha256 } from "./canonical.js";
import type { ValidationReport } from "./report.js";
import {
  validateConnectorBinding,
  validateEntityAsset,
  validateNormalizedProduct,
  validatePresentation,
  validateSceneManifest,
} from "./validation.js";

export type SceneDocument = SceneValue;
export type EntityDocument = EntityValue;
export type PresentationDocument = PresentationValue;
export type ConnectorBindingDocument = ConnectorBindingValue;
export type NormalizedProductDocument = NormalizedProductValue;

export type DeepReadonly<T> =
  T extends (...args: never[]) => unknown
    ? T
    : T extends readonly (infer Item)[]
      ? readonly DeepReadonly<Item>[]
      : T extends object
        ? { readonly [Key in keyof T]: DeepReadonly<T[Key]> }
        : T;

export interface ImmutableContractDocument<Value> {
  readonly value: DeepReadonly<Value>;
  readonly canonicalBytes: Buffer;
  readonly canonicalHash: string;
  toMutable(): Value;
}

export interface ImmutableSceneDocument
  extends ImmutableContractDocument<SceneValue> {
  readonly revision: string;
}

export type ImmutableEntityDocument = ImmutableContractDocument<EntityValue>;
export type ImmutablePresentationDocument =
  ImmutableContractDocument<PresentationValue>;
export type ImmutableConnectorBindingDocument =
  ImmutableContractDocument<ConnectorBindingValue>;
export type ImmutableNormalizedProductDocument =
  ImmutableContractDocument<NormalizedProductValue>;

export interface ContractDocumentFactory<Value, Document> {
  fromValue(value: Value | DeepReadonly<Value>): Document;
  parse(data: Uint8Array | string): Document;
}

export class ContractDocumentValidationError extends TypeError {
  public constructor(public readonly report: ValidationReport) {
    const first = report.firstError;
    super(
      first === null
        ? "Scene contract validation failed"
        : `Scene contract validation failed: ${first.code} at ${first.path}: ${first.message}`,
    );
    this.name = "ContractDocumentValidationError";
  }
}

type Validator = (value: unknown) => ValidationReport;

function deepFreeze(value: unknown): unknown {
  if (Array.isArray(value)) {
    for (const child of value) deepFreeze(child);
    return Object.freeze(value);
  }
  if (value !== null && typeof value === "object") {
    for (const child of Object.values(value)) deepFreeze(child);
    return Object.freeze(value);
  }
  return value;
}

class FrozenContractDocument<Value>
  implements ImmutableContractDocument<Value>
{
  readonly #value: DeepReadonly<Value>;
  readonly #canonicalBytes: Buffer;
  readonly #canonicalHash: string;

  public constructor(value: Value | DeepReadonly<Value>, validate: Validator) {
    const canonical = canonicalJsonBytes(value);
    const detached = parseCanonicalJson(canonical);
    const validation = validate(detached);
    if (!validation.valid) throw new ContractDocumentValidationError(validation);
    this.#value = deepFreeze(detached) as DeepReadonly<Value>;
    this.#canonicalBytes = Buffer.from(canonical);
    this.#canonicalHash = sha256(canonical);
    Object.freeze(this);
  }

  public get value(): DeepReadonly<Value> {
    return this.#value;
  }

  public get canonicalBytes(): Buffer {
    return Buffer.from(this.#canonicalBytes);
  }

  public get canonicalHash(): string {
    return this.#canonicalHash;
  }

  public toMutable(): Value {
    return parseCanonicalJson(this.#canonicalBytes) as Value;
  }
}

class FrozenSceneDocument
  extends FrozenContractDocument<SceneValue>
  implements ImmutableSceneDocument
{
  public get revision(): string {
    return this.value.revision;
  }
}

function documentFactory<Value, Document>(
  create: (value: Value | DeepReadonly<Value>) => Document,
): ContractDocumentFactory<Value, Document> {
  return Object.freeze({
    fromValue: create,
    parse: (data: Uint8Array | string) => create(parseCanonicalJson(data) as Value),
  });
}

export const SceneDocument = documentFactory<SceneValue, ImmutableSceneDocument>(
  (value) => new FrozenSceneDocument(value, validateSceneManifest),
);

export const EntityDocument = documentFactory<EntityValue, ImmutableEntityDocument>(
  (value) => new FrozenContractDocument<EntityValue>(value, validateEntityAsset),
);

export const PresentationDocument = documentFactory<
  PresentationValue,
  ImmutablePresentationDocument
>((value) => new FrozenContractDocument<PresentationValue>(value, validatePresentation));

export const ConnectorBindingDocument = documentFactory<
  ConnectorBindingValue,
  ImmutableConnectorBindingDocument
>((value) => new FrozenContractDocument<ConnectorBindingValue>(value, validateConnectorBinding));

export const NormalizedProductDocument = documentFactory<
  NormalizedProductValue,
  ImmutableNormalizedProductDocument
>((value) => new FrozenContractDocument<NormalizedProductValue>(value, validateNormalizedProduct));
